### PR TITLE
feat: access-api handling store/info for space not in db returns failure with name

### PR DESCRIPTION
**Pull request by gobengo**
Motivation:
* implement https://github.com/web3-storage/w3protocol/issues/382
  * tl;dr `space/info` now has explicit failure type for case where the ucans are valid, but the space DID is not in the db. that way when upload-api invokes `space/info` as part of its `verifyInvocation` logic, it can distinguish between an invalid invocation and a not-found space id

Unblocks:
* this test passing https://github.com/gobengo/w3protocol-test/pull/1
* rest of https://github.com/web3-storage/w3infra/issues/134#user-content-implementation

### CODE REVIEW

File: packages/access-api/src/service/index.js
Comment (hugomrdias) [inline packages/access-api/src/service/index.js:40]: we probably should start a error.js file like this https://github.com/web3-storage/ucanto/blob/main/packages/validator/src/error.js and a error.ts for error types in the client

Comment (hugomrdias) [reply packages/access-api/src/service/index.js:40]: this is gonna return a http 200 so i dont think we need the status in there, just a normal `SpaceNotFound extends Failure` with `name` and `message` props.

File: packages/access-client/src/types.ts
Comment (hugomrdias) [inline packages/access-client/src/types.ts:69]: maybe create a new errors.ts file exported from types.ts with something like https://github.com/web3-storage/ucanto/blob/main/packages/interface/src/capability.ts#L413-L416

**Review by hugomrdias — CHANGES_REQUESTED** [review]

File: packages/access-api/src/service/index.js
Comment (gobengo) [reply packages/access-api/src/service/index.js:40]: The status is in there as the indication of what kind of error this is. I don't think we should use just a specific 'message' value as the only discriminant of this error case because doing that would make it so that changing the error message string would be a backward-incompatible change from the perspective of any client trying to detect this error case. I'm intentionally trying to add a specific non-`message` [error code](https://en.wikipedia.org/wiki/Error_code) that clients can use to detect this case.

File: packages/access-client/src/types.ts
Comment (gobengo) [inline packages/access-client/src/types.ts:67]: change from status int to
```
name: 'SpaceNotFound'
```

Comment (Gozala) [inline packages/access-client/src/types.ts:67]:
```suggestion
  name: 'UnknownSpaceDID`
```

I would say ☝️, which is I think more accurate than not found. If we want to be future compatible I would go as far as do this instead.

```suggestion
  name: 'NoStorageProvider'
```

**Review by Gozala — CHANGES_REQUESTED** [review]
Provided feedback. I don't think you need another review from me.

File: packages/access-client/src/types.ts
Comment (Gozala) [reply packages/access-client/src/types.ts:67]: For what it's worth I don't mind `status` along with the `name` to categorize concrete errors into a groups with HTTP semantics, I just think it's not replacement for concrete errors.

File: packages/access-api/src/service/index.js
Comment (hugomrdias) [reply packages/access-api/src/service/index.js:40]: exactly, non-http transport should not need to understand http codes. For errors IMO either use the name as the identifier or a code nodejs style `error.code = ERR_SPACE_NOT_FOUND`

```js
// similar to ucanto
if(error.name === 'SpaceNotFound'){}

// similar to nodejs
class SpaceNotFoundError{}
SpaceNotFoundError.code = 'ERR_SPACE_NOT_FOUND'

if(error.code === SpaceNotFoundError.code) {}
```

**Comment by gobengo** [comment]
also relevant: https://github.com/web3-storage/specs/issues/18

File: packages/access-api/src/service/index.js
Comment (gobengo) [reply packages/access-api/src/service/index.js:40]: @hugomrdias Thanks, when I first responded I missed that you were encouraging a `name` and thought you were just saying to cut `status` which would make it the same as generic `Failure`. my bad

File: packages/access-client/src/types.ts
Comment (gobengo) [reply packages/access-client/src/types.ts]: I moved the failure type to an `errors.ts` file as you suggested https://github.com/web3-storage/w3protocol/pull/391/commits/8f0e5e342e858f25eb832bde4be2ea90a07175ae#diff-f4e938a2d743868395a4bb96adeecd10ce306ac53100e1e978df8d40ac33d4ef

File: packages/access-client/src/types.ts
Comment (gobengo) [reply packages/access-client/src/types.ts]: I changed it to `SpaceUnknown`

Comment (gobengo) [reply packages/access-client/src/types.ts]: I got rid of `status` and used `name` of `UnknownSpace` in https://github.com/web3-storage/w3protocol/pull/391/commits/8f0e5e342e858f25eb832bde4be2ea90a07175ae#diff-f4e938a2d743868395a4bb96adeecd10ce306ac53100e1e978df8d40ac33d4ef

File: packages/access-api/src/service/index.js
Comment (gobengo) [reply packages/access-api/src/service/index.js]: I agree it would be good to have actionable feedback in the error message. I defer to @hugomrdias on whether this is the best advice to give (accept the suggestion) or whether to deal with that in separate PR.

File: packages/access-client/src/errors.ts
Comment (hugomrdias) [inline packages/access-client/src/errors.ts:6]: we probably should extends `Failure` here same as other errors in ucanto

**Comment by gobengo** [comment]
I updated the change here based on review of @hugomrdias and @Gozala . Now the [union tag](https://en.wikipedia.org/wiki/Tagged_union) to detect this failure type is `name` equaling `SpaceUnknown` instead of anything with a `status` property. This should still be enough for the use case at hand, which is `upload-api` having enough info on the result of `space/info` that it can distinguish in `verifiyInvocation` between an invalid invocation and a valid invocation to get info for an unknown space that `access-api` does not want to (or is unable to) serve info for.

**Review by hugomrdias — APPROVED** [review]
LGTM, left a comment inline but not blocking

File: packages/access-client/src/errors.ts
Comment (gobengo) [reply packages/access-client/src/errors.ts]: I had omitted it on purpose so that access-client types didn't need to depend on ucanto (and we'd just rely on `tsc` checking the structural types), but don't have strong opinions on it either way, so I added the `extends` clause https://github.com/web3-storage/w3protocol/pull/391/commits/0b63c2e4bfd5890d7455042464bc6fd65459ffa7 .

**Pull request merged**
